### ISSUE TITLE
RSA verify failure when signature includes a Hash Function digestInfo with absent AlgorithmIdentifier.parameters

### ISSUE THREAD

**Issue by perebf** [openssl/openssl#17130]
Hello, 

RFC 8017 stablishes in "Appendix B1. Hash functions" the following: 

_The parameters field associated with these [Hash Functions] OIDs in a value of type AlgorithmIdentifier SHALL have a value of type NULL.

The parameters field associated with id-md2 and id-md5 in a value of type AlgorithmIdentifier shall have a value of type NULL.

The parameters field associated with id-sha1, id-sha224, id-sha256,  id-sha384, id-sha512, id-sha512/224, and id-sha512/256 should generally be omitted, but if present, it shall have a value of type NULL._

Nonetheless, OpenSSL > 1.1.0 does return a "bad signature" error when using SHA256 OID without parameters (picture).

Is that an expected behaviour according to Appendix B1?

![ASN codificacion](https://user-images.githubusercontent.com/74411700/143270471-da4a2669-9faa-4199-b819-75fc647a8abd.JPG)

perebf

**Comment by bernd-edlinger** [comment]
OpenSSL implements this according to RFC 3447.
@davidben any comments?

**Comment by bernd-edlinger** [comment]
From RFC 8017 I read this:
```
Exception: When formatting the DigestInfoValue in EMSA-PKCS1-v1_5
   (see Section 9.2), the parameters field associated with id-sha1,
   id-sha224, id-sha256, id-sha384, id-sha512, id-sha512/224, and
   id-sha512/256 shall have a value of type NULL.  This is to maintain
   compatibility with existing implementations and with the numeric
   information values already published for EMSA-PKCS1-v1_5, which are
   also reflected in IEEE 1363a [IEEE1363A].
```

**Comment by perebf** [comment]
@bernd-edlinger thanks for your answer, but the RFC8017 also says:

`This is to align with the definitions originally promulgated by NIST.
   For the SHA algorithms, implementations MUST accept
   AlgorithmIdentifier values both without parameters and with NULL
   parameters.`

And because it is not specified anywhere else in the RFC, it should apply to EMSA-PSS as well as to EMSA-PKCS1-v1_5.

Based on this and taking into account previous RFC definitions I don't think it makes sense for OpenSSL to not accept both implementations.

**Comment by bernd-edlinger** [comment]
But in RFC 8017. sectione 8.2.2.  Signature Verification Operation
it is specified step by step how a signature verification has to be done:
```
1.  Length checking: If the length of the signature S is not k
          octets, output "invalid signature" and stop.

      2.  RSA verification:

          a.  Convert the signature S to an integer signature
              representative s (see Section 4.2):

                 s = OS2IP (S).

          b.  Apply the RSAVP1 verification primitive (Section 5.2.2) to
              the RSA public key (n, e) and the signature representative
              s to produce an integer message representative m:

                 m = RSAVP1 ((n, e), s).

              If RSAVP1 outputs "signature representative out of range",
              output "invalid signature" and stop.

          c.  Convert the message representative m to an encoded message
              EM of length k octets (see Section 4.1):

                 EM = I2OSP (m, k).

              If I2OSP outputs "integer too large", output "invalid
              signature" and stop.

3.  EMSA-PKCS1-v1_5 encoding: Apply the EMSA-PKCS1-v1_5 encoding
          operation (Section 9.2) to the message M to produce a second
          encoded message EM' of length k octets:

             EM' = EMSA-PKCS1-V1_5-ENCODE (M, k).

          If the encoding operation outputs "message too long", output
          "message too long" and stop.  If the encoding operation
          outputs "intended encoded message length too short", output
          "RSA modulus too short" and stop.

4.  Compare the encoded message EM and the second encoded message
          EM'.  If they are the same, output "valid signature";
          otherwise, output "invalid signature".
```

Which is exactly how we do it.
So there is no need to check an alternate encoding.

So I think maybe the RFC contradicts itself, or am I missing something?

Which crypto-library is generating this alternate encoding?

**Comment by perebf** [comment]
I don't think it contradicts itself (it is really confusing though), but if we assume that "implementations MUST accept AlgorithmIdentifier values both without parameters and with NULL parameters" applies to EMSA-PKCS1-v1_5 OpenSSL should apply EMSA-PKCS1-v1_5 encoding in Step 3 using both parameters implementations and compare both outputs with message EM.
I have no specific information about which crypto-library is using this method, but we are currently struggling to vaidate some CA hierarchies because of this. 

Also, I found the following bug report from OpenSLL: [https://mta.openssl.org/pipermail/openssl-dev/2015-January/000436.html](url), which would indicate that this is something they're going to fix, but for some reason it was never included in newer versions.

**Comment by t8m** [comment]
I believe this is a confusion of the AlgorithmIdentifier as used within the PKCS1-v1.5 padding and the AlgorithmIdentifier as used outside of the signature to indicate the algorithm of the signature.

**Comment by perebf** [comment]
@t8m i'm not sure which one issue #3665 reffers to, but we are clearly reffering to the AlgorithmIdentifier as used within the PKCS1-v1.5 padding.

**Comment by t8m** [comment]
I do not think there is any ambiguity allowed for that.

**Comment by perebf** [comment]
There may not be ambiguity regarding the way the DigestInfoValue should be encoded in RFC 8017. Regardless of that it also says (or it is ambigous about it) that both encodings should be accepted by veryfing tools.
In any case, previous RFC 3447 did allow for non parameters encoding and thus, OpenSSL may be rejecting perfectly valid CA heriarchies issued prior 2016, which is odd.

**Comment by davidben** [comment]
This issue has come up before and is wrong every time.

@bernd-edlinger and @t8m are right. The use of AlgorithmIdentifier in EMSA-PKCS1-v1_5 is slightly different from the usual use. The usual use is in some structure to tell the verifier which algorithm was used. For instance, in the signature algorithm of a X.509 certificate, or the list of hashes in a PKCS#7 SignedData. There, regrettably, we ended up with a mess where two inputs are allowed.

However, in EMSA-PKCS1-v1_5, this is a domain separation mechanism deep inside the signature scheme. It is not parsed as an AlgorithmIdentifier, and indeed it's not parsed at all. There, as @bernd-edlinger notes, RFC 8017 section 8.2.2 is quite clear on how to do this. That algorithm admits no variations, and indeed it is crucial that it does not. The PKCS#1 v1.5 family of algorithms weren't designed right and are surprisingly fragile. For PKCS#1 v1.5 signatures, accepting too many classes of invalid inputs can admit forgeries. See https://www.imperialviolet.org/2014/09/26/pkcs1.html

It is a pity that we have this confusing mess of different parameterizations and variations. Alas, a lot of cryptography from this era was mis-standardized. But it is what it is. (The right way to set a signature scheme is to pick _all_ the parameters and give the _overall_ combination a name, like Ed25519 and Ed448, never reusing keys for multiple schemes. RSA is really a family of signature schemes, but we pretend it is a single one and this results in a lot of problems. Indeed AlgorithmIdentifier shouldn't have had a parameters field at all.)

> In any case, previous RFC 3447 did allow for non parameters encoding and thus, OpenSSL may be rejecting perfectly valid CA heriarchies issued prior 2016, which is odd.

Also not true. The same verification procedure exists in RFC 3447.
https://datatracker.ietf.org/doc/html/rfc3447#section-8.2.2

On the signing side, it references the encoding procedure, which references appendix A.2.4 and is similarly unambiguous about the value being a NULL ("PARAMETERS NULL").
https://datatracker.ietf.org/doc/html/rfc3447#section-9.2
https://datatracker.ietf.org/doc/html/rfc3447#appendix-A.2.4

Going further back to RFC 2437 (PKCS#1 v2.0), we see the same thing (for SHA-1 and MD5; this document predates SHA-2)
https://datatracker.ietf.org/doc/html/rfc2437#section-8.1.2
https://datatracker.ietf.org/doc/html/rfc2437#section-11.2.3

To see a different verification algorithm, we have to go all the way back to RFC 2313 (PKCS#1 v1.5). However, that document is thrice-obsolete, 23 years old, and specifies a [broken scheme](https://www.imperialviolet.org/2014/09/26/pkcs1.html), so I think it can be disregarded.
https://datatracker.ietf.org/doc/html/rfc2313#section-10.2.3

However, *even then*, the signing process specifies to use a NULL, so we are even compatible with correctly-implemented 1998-era signers. We're just using a stricter (i.e. not broken) verification process.
https://datatracker.ietf.org/doc/html/rfc2313#section-10.1.2

**Comment by perebf** [comment]
@davidben @t8m @bernd-edlinger thank you all for the clarifications and comments!

**Comment by hectorplaza** [comment]
Hello,

In the rfc 5280, in the signatureValue section it references the rfc 4055
[https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.1.3](https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.1.3)

Within this rfc, in the PKCS 1 Version 1.5 Signature Algorithm section, it states that 

>    When any of these four object identifiers appears within an
   AlgorithmIdentifier, the parameters MUST be NULL.  Implementations
   MUST accept the parameters being absent as well as present.

[https://datatracker.ietf.org/doc/html/rfc4055#section-5](https://datatracker.ietf.org/doc/html/rfc4055#section-5)

It seems that when signing a X.509 certificate, the parameteres must be NULL. However, in the specific case of verifying a certificate, both options absent and NULL could be valid. Does that make sense?

I think it is the same conclusion that was reached in the bug report @perebf mencioned before [https://mta.openssl.org/pipermail/openssl-dev/2015-January/000436.html](https://mta.openssl.org/pipermail/openssl-dev/2015-January/000436.html), where it is said that 

> OpenSSL will tolerate both an absent parameter list and a NULL one. It did before this change and still does after it.

Maybe I am missing something here, it is not clear.

**Comment by davidben** [comment]
> It seems that when signing a X.509 certificate, the parameteres must be NULL. However, in the specific case of verifying a certificate, both options absent and NULL could be valid. Does that make sense?

No, that is referring to the signatureAlgorithm field in a certificate, not the DigestInfo structure deep inside the PKCS#1 v1.5 padding. Indeed, the text you cited refers to "these four object identifiers", which are sha224WithRSAEncryption, sha256WithRSAEncryption, sha384WithRSAEncryption, and sha512WithRSAEncryption. Those are not even the same OIDs. But, even if they were, this is a different context. Please see the other text in this thread.

> OpenSSL will tolerate both an absent parameter list and a NULL one. It did before this change and still does after it.

No, that is referring to the signatureAlgorithm field in a certificate, not the DigestInfo structure deep inside the PKCS#1 v1.5 padding.

**Comment by hectorplaza** [comment]
You are absolutely right, it refers to the signatureAlgorithm. But, in the same rfc 4055 (which is also referenced in rfc 5280 in the section 4.1.1.3. signatureValue, not only in section 4.1.1.2. signatureAlgorithm), in section 2.1. One-way Hash Functions
[https://datatracker.ietf.org/doc/html/rfc4055#section-2.1](https://datatracker.ietf.org/doc/html/rfc4055#section-2.1) 
states the following

>  All implementations MUST accept both NULL and absent parameters as
   legal and equivalent encodings.

Does it only applies to RSASSA-PSS and RSAES-OAEP?

Thank you for the clarifications

**Comment by bernd-edlinger** [comment]
Sorry, we are not lawyers, we are cryptographers.

**Comment by kaduk** [comment]
RFC 4055 applies only to RSASSA-PSS and RSAES-OAEP, yes
